basement/Basement/Cast.hs

```diff
@@ -0,0 +1,151 @@
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE FlexibleInstances     #-}
+{-# LANGUAGE ScopedTypeVariables   #-}
+{-# LANGUAGE ConstraintKinds       #-}
+{-# LANGUAGE CPP                   #-}
+{-# LANGUAGE MagicHash             #-}
+{-# LANGUAGE UndecidableInstances  #-}
+-- |
+-- Module      : Basement.Cast
+-- License     : BSD-style
+-- Maintainer  : Haskell Foundation
+--
+module Basement.Cast
+    ( Cast(..)
+    ) where
+
+import           Basement.Compat.Base
+
+import           GHC.Types
+import           GHC.Prim
+import           GHC.Int
+import           GHC.Word
+import           Basement.Numerical.Number
+import           Basement.Numerical.Conversion
+import qualified Basement.Block.Base as Block
+import qualified Basement.BoxedArray as BoxArray
+import qualified Basement.UArray as UArray
+import qualified Basement.String as String
+import qualified Basement.Types.AsciiString as AsciiString
+import           Basement.Types.Word128 (Word128(..))
+import           Basement.Types.Word256 (Word256(..))
+import qualified Basement.Types.Word128 as Word128
+import qualified Basement.Types.Word256 as Word256
+import           Basement.These
+import           Basement.PrimType (PrimType)
+import           Basement.Types.OffsetSize
+import           Basement.Compat.Natural
+import qualified Prelude (fromIntegral)
+
+-- | `Cast` an object of type a to b.
+--
+-- Do not add instance of this class if the source type is not of the same
+-- size of the destination type. Also keep in mind this is casting a value
```

Comment:
Some of the instances directly disagree with this. `Word128` -> `Word8` is not the same size. I think I agree with the documentation of Cast.